### PR TITLE
Fix and clarify notes on result ordering

### PR DESCRIPTION
**Pull request by shwina**
I noticed when answering #13254 that the code example in this section of our documentation was incorrect and the text itself could use some improving.

### CODE REVIEW

File: docs/cudf/source/user_guide/pandas-comparison.md [hunk 1]
```diff
@@ -82,9 +82,16 @@ using `.from_arrow()` or `.from_pandas()`.
 
 ## Result ordering
 
-By default, `join` (or `merge`) and `groupby` operations in cuDF
-do *not* guarantee output ordering.
-Compare the results obtained from Pandas and cuDF below:
+In Pandas, `join` and `groupby` operations provide certain guarantees
+about the order of rows in the result returned.  In Pandas `join`, the
```
Comment (wence-) [inline docs/cudf/source/user_guide/pandas-comparison.md:86]:
```suggestion
about the order of rows in the result returned.  In a Pandas `join`, the
```

File: docs/cudf/source/user_guide/pandas-comparison.md [hunk 2]
```diff
@@ -82,9 +82,16 @@ using `.from_arrow()` or `.from_pandas()`.
...
+order of join keys are either preserved or sorted lexicographically by
```
Comment (wence-) [inline docs/cudf/source/user_guide/pandas-comparison.md:87]:
```suggestion
order of join keys is either preserved or sorted lexicographically by
```
(subject of the sentence is "order" not "join keys")

But I think I don't quite understand the explanation of this restriction (how can ordering be either preserved or sorted by default for a join key?).

File: docs/cudf/source/user_guide/pandas-comparison.md [hunk 3]
```diff
@@ -82,9 +82,16 @@ using `.from_arrow()` or `.from_pandas()`.
...
+default.  `groupby` sorts the group keys, and preserves the order of
+rows within each group. In some cases, disabling this option in Pandas
+can yield better performance.
+
+By contrast, cuDF's default behavior is to return rows in a
+non-deterministic order to maximize performance.  Compare the results
+obtained from Pandas and cuDF below:
```
Comment (wence-) [inline docs/cudf/source/user_guide/pandas-comparison.md:94]: Maybe note that for a groupby _within_ a group the ordering of the input is preserved?

File: docs/cudf/source/user_guide/pandas-comparison.md [hunk 4]
```diff
@@ -107,10 +114,11 @@ Compare the results obtained from Pandas and cuDF below:
  10  640.00
 ```
 
-To match Pandas behavior, you must explicitly pass `sort=True`:
+If you require that results be returned in a predictable order, you
+must pass the `sort=True` option explicitly:
```
Comment (wence-) [inline docs/cudf/source/user_guide/pandas-comparison.md:118]: Is it worth mentioning about the index being useful to write ordering-independent code?

**Review by wence- — COMMENTED** [review]
Minor nitpicks

File: docs/cudf/source/user_guide/pandas-comparison.md (hunk 3 above)
Comment (shwina) [reply docs/cudf/source/user_guide/pandas-comparison.md:94]: Do we _need_ to make that guarantee? Internally, we do use a stable sort when doing a sort-based groupby but can we leave that as an implementation detail?

File: docs/cudf/source/user_guide/pandas-comparison.md (hunk 4 above)
Comment (shwina) [reply docs/cudf/source/user_guide/pandas-comparison.md]: Done!

File: docs/cudf/source/user_guide/pandas-comparison.md (hunk 3 above)
Comment (wence-) [reply docs/cudf/source/user_guide/pandas-comparison.md:94]: Oh probably not, but I don't know the details of what we guarantee as an API statement, and what we provide as an implementation detail. So let us leave it as is here.

File: docs/cudf/source/user_guide/pandas-comparison.md (hunk 2 above)
Comment (wence-) [reply docs/cudf/source/user_guide/pandas-comparison.md:87]: OK, let us therefore write something like:

> In a Pandas `join`, the order of join keys is (depending on the particular style of join being performed) either preserved or sorted lexicographically.

**Comment by vyasr** [comment]
@shwina @wence- the docs have been updated from underneath this PR, so it needs some conflict resolution. Do you want to fix up the wording to something that you both like?

File: docs/cudf/source/user_guide/pandas-comparison.md (hunk 2 above)
Comment (vyasr) [reply docs/cudf/source/user_guide/pandas-comparison.md]: Added this.

**Review by wence- — APPROVED** [review]

**Pull request merged**